prow/config/cache.go

```diff
@@ -0,0 +1,388 @@
+
+	prowYAML, err := GetProwYAMLFromCache(c.ProwYAMLCache, c.getProwYAMLNoDefault, gc, identifier, baseSHAGetter, headSHAGetters...)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := DefaultAndValidateProwYAML(c, prowYAML, identifier); err != nil {
+		return nil, err
+	}
+
+	return append(c.GetPresubmitsStatic(identifier), prowYAML.Presubmits...), nil
+}
+
+// GetPostsubmitsFromCache is like GetPostsubmits, but attempts to use a cache
+// lookup to get the *ProwYAML value (cache hit), instead of computing it from
+// scratch (cache miss).
+func (c *Config) GetPostsubmitsFromCache(gc git.ClientFactory, identifier string, baseSHAGetter RefGetter, headSHAGetters ...RefGetter) ([]Postsubmit, error) {
+
+	prowYAML, err := GetProwYAMLFromCache(c.ProwYAMLCache, c.getProwYAMLNoDefault, gc, identifier, baseSHAGetter, headSHAGetters...)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := DefaultAndValidateProwYAML(c, prowYAML, identifier); err != nil {
+		return nil, err
+	}
+
+	return append(c.GetPostsubmitsStatic(identifier), prowYAML.Postsubmits...), nil
+}
+
+// GetProwYAMLFromCache uses ProwYAMLCache to first try to perform a lookup of
+// previously-calculated ProwYAML objects. The 'valConstructor' function is
+// taken as an argument to make it easier to test this function. This way, unit
+// tests can just provide its own function for constructing a ProwYAML object
+// (instead of needing to create an actual Git repo, etc.).
+func GetProwYAMLFromCache(
+	prowYAMLCache *ProwYAMLCache,
+	valConstructorHelper func(git.ClientFactory, string, RefGetter, ...RefGetter) (*ProwYAML, error),
+	gc git.ClientFactory,
+	identifier string,
+	baseSHAGetter RefGetter,
+	headSHAGetters ...RefGetter) (*ProwYAML, error) {
+
+	keyConstructor := func() (CacheKey, error) {
+		kp, err := MakeCacheKeyParts(identifier, baseSHAGetter, headSHAGetters...)
+		if err != nil {
+			return CacheKey(""), err
+		}
+
+		return MakeCacheKey(kp)
+	}
+
+	// The point of valConstructor is to allow us to mock this expensive value
+	// constructor call in tests (and, e.g., avoid having to do things like
+	// going over the network or dealing with an actual Git client, etc.).
+	valConstructor := func() (interface{}, error) {
+		return valConstructorHelper(gc, identifier, baseSHAGetter, headSHAGetters...)
+	}
+
+	val, err := prowYAMLCache.GetOrAdd(keyConstructor, valConstructor)
+	if err != nil {
+		return nil, err
+	}
+
+	prowYAML, ok := val.(*ProwYAML)
+	if ok {
+		return prowYAML, err
+	}
+
+	// Somehow, the value retrieved with GetFromCache has a malformed type. This
+	// can happen if some other function modified the cache. Ultimately, this is
+	// a price we pay for using a cache library that uses "interface{}" for the
+	// type of its items. In this case, we log a warning and return an error.
+	err = fmt.Errorf("cache value type error: expected value type '*config.ProwYAML', got '%T'", val)
+	logrus.Warn(err)
```

Comment:
Is this recoverable error? if not, this log should be on error level

Reply:
> Is this recoverable error?

To recover we would have to delete the bad key/value pair, but also perform some sort of retry. I think the engineering effort for handling this particular recovery is not worth it, given how it most likely boils down to programmer error.

> if not, this log should be on error level

done

> Maybe also indicate this is a programmer error.

done